Inconsistent behavior when passing empty Points to MultiPoint

The [documentation on `MultiPoint`](https://shapely.readthedocs.io/en/latest/manual.html#MultiPoint) says that I can pass to it a sequence of `Point`s. But it doesn't warn us against passing empty `Point`s. Interestingly enough, if I pass an empty `Point` as the first element, it will throw an `AssertionError`:
```python
>>> from shapely.geometry import Point, MultiPoint
>>> MultiPoint([Point()])
```
```none
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-30-8942eaebd9a2> in <module>
----> 1 MultiPoint([Point()])

~\Miniconda3\lib\site-packages\shapely\geometry\multipoint.py in __init__(self, points)
     56             pass
     57         else:
---> 58             self._geom, self._ndim = geos_multipoint_from_py(points)
     59 
     60     def shape_factory(self, *args):

~\Miniconda3\lib\site-packages\shapely\geometry\multipoint.py in geos_multipoint_from_py(ob)
    162     except TypeError:
    163         n = ob[0]._ndim
--> 164     assert n == 2 or n == 3
    165 
    166     # Array of pointers to point geometries

AssertionError: 
```
```python
>>> MultiPoint([Point(), Point(1, 2)])
```
```none
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-31-08a0b83c889a> in <module>
----> 1 MultiPoint([Point(), Point(1, 2)])

~\Miniconda3\lib\site-packages\shapely\geometry\multipoint.py in __init__(self, points)
     56             pass
     57         else:
---> 58             self._geom, self._ndim = geos_multipoint_from_py(points)
     59 
     60     def shape_factory(self, *args):

~\Miniconda3\lib\site-packages\shapely\geometry\multipoint.py in geos_multipoint_from_py(ob)
    162     except TypeError:
    163         n = ob[0]._ndim
--> 164     assert n == 2 or n == 3
    165 
    166     # Array of pointers to point geometries

AssertionError: 
```
But if the empty `Point` goes after a nonempty `Point`, then it works fine:
```python
>>> MultiPoint([Point(1, 2), Point()])
>>> # no error here
```

But the resulting object will throw an exception after trying to get its wkt:
```python
>>> MultiPoint([Point(1, 2), Point()]).wkt
```
```none
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-38-c645b8b312e6> in <module>
----> 1 MultiPoint([Point(1, 2), Point()]).wkt

~\Miniconda3\lib\site-packages\shapely\geometry\base.py in wkt(self, **kw)
    365     def wkt(self, **kw):
    366         """WKT representation of the geometry"""
--> 367         return WKTWriter(lgeos, **kw).write(self)
    368 
    369     @property

~\Miniconda3\lib\site-packages\shapely\geos.py in write(self, geom)
    361         if geom is None or geom._geom is None:
    362             raise ValueError("Null geometry supports no operations")
--> 363         result = self._lgeos.GEOSWKTWriter_write(self._writer, geom._geom)
    364         text = string_at(result)
    365         lgeos.GEOSFree(result)

OSError: exception: access violation reading 0x0000000000000000
```
I assume there needs to be a guard against a presence of empty points in the passed sequence, and the docs should be updated too.

---
Shapely version: 1.6.4.post1, installed from conda.

--- a/shapely/errors.py
+++ b/shapely/errors.py
@@ -31,3 +31,7 @@ class TopologicalError(ShapelyError):
 
 class PredicateError(ShapelyError):
     """A geometric predicate has failed to return True/False."""
+
+
+class EmptyPartError(ShapelyError):
+    """An error signifying an empty part was encountered when creating a multi-part."""

--- a/shapely/geometry/multilinestring.py
+++ b/shapely/geometry/multilinestring.py
@@ -132,6 +132,10 @@ def geos_multilinestring_from_py(ob):
     # add to coordinate sequence
     for l in range(L):
         geom, ndims = linestring.geos_linestring_from_py(obs[l])
+
+        if lgeos.GEOSisEmpty(geom):
+            raise ValueError("Can't create MultiLineString with empty component")
+
         subs[l] = cast(geom, c_void_p)
             
     return (lgeos.GEOSGeom_createCollection(5, subs, L), N)

--- a/shapely/geometry/multilinestring.py
+++ b/shapely/geometry/multilinestring.py
@@ -8,6 +8,7 @@ if sys.version_info[0] < 3:
 
 from ctypes import c_void_p, cast
 
+from shapely.errors import EmptyPartError, DimensionError
 from shapely.geos import lgeos
 from shapely.geometry.base import BaseMultipartGeometry, geos_geom_from_py
 from shapely.geometry import linestring
@@ -134,7 +135,7 @@ def geos_multilinestring_from_py(ob):
         geom, ndims = linestring.geos_linestring_from_py(obs[l])
 
         if lgeos.GEOSisEmpty(geom):
-            raise ValueError("Can't create MultiLineString with empty component")
+            raise EmptyPartError("Can't create MultiLineString with empty component")
 
         subs[l] = cast(geom, c_void_p)
             

--- a/shapely/geometry/multilinestring.py
+++ b/shapely/geometry/multilinestring.py
@@ -8,7 +8,7 @@ if sys.version_info[0] < 3:
 
 from ctypes import c_void_p, cast
 
-from shapely.errors import EmptyPartError, DimensionError
+from shapely.errors import EmptyPartError
 from shapely.geos import lgeos
 from shapely.geometry.base import BaseMultipartGeometry, geos_geom_from_py
 from shapely.geometry import linestring

--- a/shapely/geometry/multipoint.py
+++ b/shapely/geometry/multipoint.py
@@ -8,6 +8,7 @@ if sys.version_info[0] < 3:
 
 from ctypes import byref, c_double, c_void_p, cast
 
+from shapely.errors import EmptyPartError, DimensionError
 from shapely.geos import lgeos
 from shapely.geometry.base import (
     BaseMultipartGeometry, exceptNull, geos_geom_from_py)
@@ -171,7 +172,7 @@ def geos_multipoint_from_py(ob):
         geom, ndims = point.geos_point_from_py(coords)
 
         if lgeos.GEOSisEmpty(geom):
-            raise ValueError("Can't create MultiPoint with empty component")
+            raise EmptyPartError("Can't create MultiPoint with empty component")
 
         subs[i] = cast(geom, c_void_p)
 

--- a/shapely/geometry/multipoint.py
+++ b/shapely/geometry/multipoint.py
@@ -8,7 +8,7 @@ if sys.version_info[0] < 3:
 
 from ctypes import byref, c_double, c_void_p, cast
 
-from shapely.errors import EmptyPartError, DimensionError
+from shapely.errors import EmptyPartError
 from shapely.geos import lgeos
 from shapely.geometry.base import (
     BaseMultipartGeometry, exceptNull, geos_geom_from_py)

--- a/shapely/geometry/multipoint.py
+++ b/shapely/geometry/multipoint.py
@@ -169,6 +169,10 @@ def geos_multipoint_from_py(ob):
     for i in range(m):
         coords = ob[i]
         geom, ndims = point.geos_point_from_py(coords)
+
+        if lgeos.GEOSisEmpty(geom):
+            raise ValueError("Can't create MultiPoint with empty component")
+
         subs[i] = cast(geom, c_void_p)
 
     return lgeos.GEOSGeom_createCollection(4, subs, m), n

--- a/tests/test_multilinestring.py
+++ b/tests/test_multilinestring.py
@@ -1,5 +1,6 @@
 from . import unittest, numpy, test_int_types
 from .test_multi import MultiGeometryTestCase
+from shapely.errors import EmptyPartError
 from shapely.geos import lgeos
 from shapely.geometry import LineString, MultiLineString, asMultiLineString
 from shapely.geometry.base import dump_coords
@@ -80,7 +81,7 @@ class MultiLineStringTestCase(MultiGeometryTestCase):
         self.subgeom_access_test(MultiLineString, [line0, line1])
 
     def test_create_multi_with_empty_component(self):
-        with self.assertRaises(ValueError) as exc:
+        with self.assertRaises(EmptyPartError) as exc:
             wkt = MultiLineString([
                 LineString([(0, 0), (1, 1), (2, 2)]),
                 LineString()

--- a/tests/test_multilinestring.py
+++ b/tests/test_multilinestring.py
@@ -87,7 +87,7 @@ class MultiLineStringTestCase(MultiGeometryTestCase):
                 LineString()
             ]).wkt
 
-        self.assertRegex(str(exc.exception), "Can't create MultiLineString with empty component")
+        self.assertEqual(str(exc.exception), "Can't create MultiLineString with empty component")
 
 
 def test_suite():

--- a/tests/test_multilinestring.py
+++ b/tests/test_multilinestring.py
@@ -79,6 +79,15 @@ class MultiLineStringTestCase(MultiGeometryTestCase):
         line1 = LineString([(4.0, 5.0), (6.0, 7.0)])
         self.subgeom_access_test(MultiLineString, [line0, line1])
 
+    def test_create_multi_with_empty_component(self):
+        with self.assertRaises(ValueError) as exc:
+            wkt = MultiLineString([
+                LineString([(0, 0), (1, 1), (2, 2)]),
+                LineString()
+            ]).wkt
+
+        self.assertRegex(str(exc.exception), "Can't create MultiLineString with empty component")
+
 
 def test_suite():
     return unittest.TestLoader().loadTestsFromTestCase(MultiLineStringTestCase)

--- a/tests/test_multipoint.py
+++ b/tests/test_multipoint.py
@@ -1,5 +1,6 @@
 from . import unittest, numpy, test_int_types
 from .test_multi import MultiGeometryTestCase
+from shapely.errors import EmptyPartError
 from shapely.geometry import Point, MultiPoint, asMultiPoint
 from shapely.geometry.base import dump_coords
 
@@ -75,7 +76,7 @@ class MultiPointTestCase(MultiGeometryTestCase):
         self.subgeom_access_test(MultiPoint, [p0, p1])
 
     def test_create_multi_with_empty_component(self):
-        with self.assertRaises(ValueError) as exc:
+        with self.assertRaises(EmptyPartError) as exc:
             wkt = MultiPoint([Point(0, 0), Point()]).wkt
 
         self.assertRegex(str(exc.exception), "Can't create MultiPoint with empty component")

--- a/tests/test_multipoint.py
+++ b/tests/test_multipoint.py
@@ -79,7 +79,7 @@ class MultiPointTestCase(MultiGeometryTestCase):
         with self.assertRaises(EmptyPartError) as exc:
             wkt = MultiPoint([Point(0, 0), Point()]).wkt
 
-        self.assertRegex(str(exc.exception), "Can't create MultiPoint with empty component")
+        self.assertEqual(str(exc.exception), "Can't create MultiPoint with empty component")
 
 
 def test_suite():

--- a/tests/test_multipoint.py
+++ b/tests/test_multipoint.py
@@ -1,4 +1,4 @@
-from . import unittest, numpy, test_int_types
+from . import unittest, numpy
 from .test_multi import MultiGeometryTestCase
 from shapely.geometry import Point, MultiPoint, asMultiPoint
 from shapely.geometry.base import dump_coords
@@ -73,6 +73,13 @@ class MultiPointTestCase(MultiGeometryTestCase):
         p0 = Point(1.0, 2.0)
         p1 = Point(3.0, 4.0)
         self.subgeom_access_test(MultiPoint, [p0, p1])
+
+    def test_create_multi_with_empty_component(self):
+        with self.assertRaises(ValueError) as exc:
+            wkt = MultiPoint([Point(0, 0), Point()]).wkt
+
+        self.assertRegex(str(exc.exception), "Can't create MultiPoint with empty component")
+
 
 def test_suite():
     return unittest.TestLoader().loadTestsFromTestCase(MultiPointTestCase)

--- a/tests/test_multipoint.py
+++ b/tests/test_multipoint.py
@@ -1,4 +1,4 @@
-from . import unittest, numpy
+from . import unittest, numpy, test_int_types
 from .test_multi import MultiGeometryTestCase
 from shapely.geometry import Point, MultiPoint, asMultiPoint
 from shapely.geometry.base import dump_coords

